File: typescript/tests/ShoppingCart.test.ts
import {afterEach, describe, expect, it, vi} from 'vitest'

import {ShoppingCart} from '../src/ShoppingCart';
import {Stock} from '../src/Stock';
import {CurrentUser} from '../src/CurrentUser';
import {PriceCalculator} from "../src/PriceCalculator";
import {ShippingCalculator} from "../src/ShippingCalculator";
import {Article} from "../src/Article";
import {CustomerStatus} from "../src/CustomerStatus";

describe('ShoppingCart', () => {
    afterEach(() => {
        vi.restoreAllMocks()
    })

    it('add', () => {
        const stock = new Stock()
        const spyStock = vi.spyOn(stock, 'availableUnits')
        spyStock.mockImplementationOnce(() => 2).mockImplementationOnce(() => 3)
        const currentUser1 = new CurrentUser(CustomerStatus.GOLD);
        let priceCalculator = new PriceCalculator();
        const spyPriceCalculator = vi.spyOn(priceCalculator, 'calculatePrice')
        spyPriceCalculator.mockImplementationOnce(() => 9.95).mockImplementationOnce(() => 7.5)
        let shippingCalculator = new ShippingCalculator();
        const spyShippingCalculator = vi.spyOn(shippingCalculator, 'calculateShipping')
        spyShippingCalculator.mockImplementationOnce(() => 3.5).mockImplementationOnce(() => 3.5)
        const article1 = new Article('', '', '', '');
        const article2 = new Article('', '', '', '');
        const shoppingCart = new ShoppingCart(stock, currentUser1, priceCalculator, shippingCalculator);
        shoppingCart.add(article1, 1)
        shoppingCart.add(article2, 3)
        expect(spyStock).toHaveBeenCalledWith(article1)
        expect(spyStock).toHaveBeenCalledWith(article2)
        expect(spyPriceCalculator).toHaveBeenCalledWith(article1, CustomerStatus.GOLD)
        expect(spyPriceCalculator).toHaveBeenCalledWith(article2, CustomerStatus.GOLD)
        expect(spyShippingCalculator).toHaveBeenCalledWith(9.95)
        expect(spyShippingCalculator).toHaveBeenCalledWith(32.45)
        expect(shoppingCart.numberOfItems()).toBe(2)
        expect(shoppingCart.getItems()).toBeDefined()
        expect(shoppingCart.getItems()[0]).toBeDefined()
        expect(shoppingCart.getItems()[0].quantity).toBe(1)
        expect(shoppingCart.getItems()[0].amount).toBe(9.95)
        expect(shoppingCart.getItems()[1]).toBeDefined()
        expect(shoppingCart.getItems()[1].quantity).toBe(3)
        expect(shoppingCart.getItems()[1].amount).toBe(22.5)
        expect(shoppingCart.getSubtotalAmount()).toBe(32.45)
        expect(shoppingCart.getShippingAmount()).toBe(3.5)
        expect(shoppingCart.getTotalAmount()).toBe(35.95)
    })
});

File: java/src/test/java/com/innoq/automatedtesting/samples/shoppingcart/ShoppingCartTest.java
package com.innoq.automatedtesting.samples.shoppingcart;

import org.junit.jupiter.api.Test;

import java.math.BigDecimal;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.mockito.Mockito.*;

public class ShoppingCartTest {

    @Test
    public void add() throws InsufficientUnitsInStockException {
        var stock = mock(Stock.class);
        var article1 = mock(Article.class);
        when(stock.availableUnits(article1)).thenReturn(2);
        var currentUser = mock(CurrentUser.class);
        when(currentUser.getCustomerStatus()).thenReturn(CustomerStatus.GOLD);
        var priceCalculator = mock(PriceCalculator.class);
        when(priceCalculator.calculatePrice(article1, CustomerStatus.GOLD)).thenReturn(BigDecimal.valueOf(9.95));
        var shippingCalculator = mock(ShippingCalculator.class);
        when(shippingCalculator.calculateShipping(BigDecimal.valueOf(9.95))).thenReturn(BigDecimal.valueOf(3.5));
        var shoppingCart = new ShoppingCart(stock, currentUser, priceCalculator, shippingCalculator);
        shoppingCart.add(article1, 1);
        var article2 = mock(Article.class);
        when(stock.availableUnits(article2)).thenReturn(3);
        when(priceCalculator.calculatePrice(article2, CustomerStatus.GOLD)).thenReturn(BigDecimal.valueOf(7.5));
        when(shippingCalculator.calculateShipping(BigDecimal.valueOf(32.45))).thenReturn(BigDecimal.valueOf(3.5));
        shoppingCart.add(article2, 3);
        verify(stock).availableUnits(article1);
        verify(stock).availableUnits(article2);
        verify(currentUser, times(2)).getCustomerStatus();
        verify(priceCalculator).calculatePrice(article1, CustomerStatus.GOLD);
        verify(priceCalculator).calculatePrice(article2, CustomerStatus.GOLD);
        verify(shippingCalculator).calculateShipping(BigDecimal.valueOf(9.95));
        verify(shippingCalculator).calculateShipping(BigDecimal.valueOf(32.45));
        assertNotNull(shoppingCart.items());
        assertEquals(2, shoppingCart.items().size());
        assertNotNull(shoppingCart.items().get(0));
        assertEquals(1, shoppingCart.items().get(0).quantity());
        assertEquals(BigDecimal.valueOf(9.95), shoppingCart.items().get(0).amount());
        assertNotNull(shoppingCart.items().get(1));
        assertEquals(3, shoppingCart.items().get(1).quantity());
        assertEquals(BigDecimal.valueOf(22.50), shoppingCart.items().get(1).amount());
        assertEquals(BigDecimal.valueOf(32.45), shoppingCart.subtotalAmount());
        assertEquals(BigDecimal.valueOf(3.5), shoppingCart.shippingAmount());
        assertEquals(BigDecimal.valueOf(35.95), shoppingCart.totalAmount());
    }
}
